bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperAdmin.java

```diff
@@ -739,19 +684,40 @@ public void openComplete(int rc, final LedgerHandle lh, Object ctx) {
                         asyncOpenLedger(lId, new OpenCallback() {
                             @Override
                             public void openComplete(int newrc, final LedgerHandle newlh, Object newctx) {
-                                if (newrc != Code.OK.intValue()) {
+                                if (newrc != BKException.Code.OK) {
                                     LOG.error("BK error close ledger: " + lId, BKException.create(newrc));
-                                    ledgerIterCb.processResult(newrc, null, null);
+                                    finalLedgerIterCb.processResult(newrc, null, null);
                                     return;
                                 }
-                                // do recovery
-                                recoverLedger(bookieSrc, lId, ledgerIterCb, availableBookies);
+                                bkc.mainWorkerPool.submit(() -> {
+                                    // do recovery
+                                    recoverLedger(bookiesSrc, lId, dryrun, skipOpenLedgers, finalLedgerIterCb);
+                                });
                             }
                         }, null);
                         return;
                     }
                 }
 
+                final AsyncCallback.VoidCallback ledgerIterCb = new AsyncCallback.VoidCallback() {
```

Comment:
I don't think it is worth a util method, because the logging info is more recover related. we can improve it using lambda and completable future in future.

Reply:
The reason I suggested the utility message is that when I came to this in the review, I was like "why is he adding this wrapper here". The logging within the callback doesn't clarify why.

Reply:
okay. I would leave it know and have #655 clean this up later. is that okay?

Reply:
no problem with me.